src/url-loader/indirect-url-resolver.ts

```diff
@@ -0,0 +1,122 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {FileRelativeUrlBrand, PackageRelativeUrl} from '../index';
+import {FileRelativeUrl, ResolvedUrl} from '../model/url';
+
+import {FsUrlResolver} from './fs-url-resolver';
+
+type RuntimeUrl = string&RuntimeUrlBrand;
+
+
+// Declare these as classes rather than interfaces so that the properties
+// can be private.
+declare class RuntimeUrlBrand extends FileRelativeUrlBrand {
+  private RuntimeUrlBrand: never;
+}
+
+/**
+ * A URL resolver for very large codebases where source files map in an
```

Comment:
"very large codebases" -> "monorepo" might be more descriptive. I think the key is that this supports many packages in one arbitrary FS layout, rather than requiring dependencies to be installed in a specific place according to a package manager.

Reply:
Also, seems like `PackageUrlResolver` is a special case of this, and that it might be easier to think about if we had a `BowerUrlResolver` and `NpmUrlResolver`, rather than configuration in `PackageUrlResolver`

Reply:
Agreed, but would like to punt this